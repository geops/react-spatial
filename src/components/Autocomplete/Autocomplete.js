import React, { PureComponent } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import SearchInput from '../SearchInput';
import List from '../List';

const propTypes = {
  /**
   * Content for submit button.
   */
  button: PropTypes.any,

  /**
   * Value of the input.
   */
  value: PropTypes.string,

  /**
   * Placeholder of the input.
   */
  placeholder: PropTypes.string,

  /**
   * Items to display on input change.
   */
  items: PropTypes.array,

  /**
   * Items always display in suggestions list.
   */
  defaultItems: PropTypes.array,

  /**
   * Function that render the title display on top the default items list.
   */
  renderTitle: PropTypes.func,

  /**
   * Function that render an item.
   */
  renderItem: PropTypes.func,

  /**
   * Function triggered on input change, focus, and on button click.
   */
  onChange: PropTypes.func,

  /**
   * Function triggered on item selection and on click outside the autocomplete
   * element.
   */
  onSelect: PropTypes.func,

  /**
   * Function triggered when the autocomplete component (input, buttons and list) get the focus.
   * element.
   */
  onFocus: PropTypes.func,

  /**
   * Function triggered when the autocomplete component (input, buttons and list) loose the focus.
   * element.
   */
  onBlur: PropTypes.func,

  /**
   * Get the key for each item used when react creates the list.
   */
  getItemKey: PropTypes.func,

  /**
   * CSS class added to container.
   */
  className: PropTypes.string,

  /**
   * CSS class added to container of the results list.
   */
  classNameResults: PropTypes.string,
};

const defaultProps = {
  button: 'submit',
  value: '',
  placeholder: '',
  items: [],
  defaultItems: [],
  renderTitle: t => t,
  renderItem: i => i,
  getItemKey: k => k,
  onBlur: () => {},
  onChange: () => {},
  onFocus: () => {},
  onSelect: () => {},
  className: 'tm-autocomplete',
  classNameResults: 'tm-autocomplete-results',
};

/**
 * This component displays a search input with a list of suggestions.
 */
class Autocomplete extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      showList: false,
      refList: null,
      focus: false,
    };
    this.onDocClick = this.onDocClick.bind(this);
    this.ref = null;
    this.refSearchInput = null;
  }

  componentDidMount() {
    // Close the list when clicking outside the list or the input
    document.addEventListener('click', this.onDocClick);
  }

  componentDidUpdate(prevProps, prevState) {
    // Close the list when clicking outside the list or the input
    document.addEventListener('click', this.onDocClick);

    // Trigger onFocus and onBlur event
    const { focus } = this.state;
    const { onBlur, onFocus } = this.props;
    if (prevState.focus && !focus) {
      onBlur();
    } else if (!prevState.focus && focus) {
      onFocus();
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.onDocClick);
  }

  onDocClick(e) {
    // If the click comes from an element of Autocomplete, don't close the list.
    if (this.ref && this.ref.contains(e.target)) {
      return;
    }
    this.setState({ showList: false, focus: false });
  }

  onChange(evt, val) {
    const { onChange } = this.props;
    this.setState({ showList: true, focus: true });
    onChange(val);
  }

  onFocus() {
    this.setState({ showList: true, focus: true });
  }

  onBlurInput(evt, lastKeyPress) {
    if (lastKeyPress && lastKeyPress.which !== 40) {
      this.setState({ showList: false, focus: false });
    }
  }

  onKeyPress(evt) {
    const { refList } = this.state;
    if (evt.which === 40 && refList && refList.querySelector('.tm-list-item')) {
      // Give focus to the first element of the results
      // eslint-disable-next-line react/no-find-dom-node
      refList.querySelector('.tm-list-item').focus();
    }
  }

  onKeyPressItem(evt) {
    if (evt.which !== 40 && evt.which !== 38) {
      return;
    }
    let delta = 1;
    if (evt.which === 38) {
      delta = -1;
    }
    const { refList } = this.state;
    const liFocused = document.activeElement;
    const lis = refList.querySelectorAll('.tm-list-item');
    const idxItemFocused = Array.prototype.slice.call(lis).indexOf(liFocused);
    let nextIndex = idxItemFocused + delta;
    if (nextIndex < 0) {
      // Move focus to input search
      // eslint-disable-next-line react/no-find-dom-node
      ReactDOM.findDOMNode(this.refSearchInput)
        .querySelector('input')
        .focus();
      nextIndex = lis.length - 1;
      return;
    }
    if (nextIndex === lis.length) {
      // Move focus to the beginning of the list
      nextIndex = 0;
    }
    lis[nextIndex].focus();
  }

  onSelect(evt, item) {
    const { onSelect } = this.props;
    onSelect(item);
    this.setState({ showList: false, focus: false });
  }

  render() {
    const {
      button,
      value,
      placeholder,
      items,
      defaultItems,
      renderTitle,
      renderItem,
      getItemKey,
      className,
      classNameResults,
    } = this.props;
    const { showList, refList } = this.state;
    const display =
      !showList || (!items.length && !defaultItems.length) ? 'none' : null;
    const hr = items.length && defaultItems.length ? <hr /> : null;

    return (
      <div
        className={className}
        ref={node => {
          this.ref = node;
        }}
      >
        <SearchInput
          button={button}
          value={value}
          placeholder={placeholder}
          onChange={(e, val) => this.onChange(e, val)}
          onFocus={e => this.onFocus(e)}
          onKeyPress={e => this.onKeyPress(e)}
          onBlurInput={(e, keyPress) => this.onBlurInput(e, keyPress)}
          onClickSearchButton={() => {
            if (showList) {
              this.setState({ showList: false, focus: false });
            }
          }}
          ref={node => {
            this.refSearchInput = node;
          }}
        />
        <div
          ref={node => {
            if (node && !refList) {
              this.setState({ refList: node });
            }
          }}
          style={
            display && {
              display,
            }
          }
          className={classNameResults}
        >
          <List
            items={items}
            renderItem={item => renderItem(item)}
            getItemKey={item => getItemKey(item)}
            onSelect={(e, item) => {
              this.onSelect(e, item);
            }}
            onKeyDownItem={e => {
              this.onKeyPressItem(e);
            }}
          />
          {hr}
          {renderTitle()}
          <List
            items={defaultItems}
            renderItem={item => renderItem(item)}
            getItemKey={item => getItemKey(item)}
            onSelect={(e, item) => {
              this.onSelect(e, item);
            }}
            onKeyDownItem={e => {
              this.onKeyPressItem(e);
            }}
          />
        </div>
      </div>
    );
  }
}

Autocomplete.propTypes = propTypes;
Autocomplete.defaultProps = defaultProps;

export default Autocomplete;
