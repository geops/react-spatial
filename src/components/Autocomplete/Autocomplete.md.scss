.tm-autocomplete-example {
  position: relative;
  width: 420px;
  height: 44px;

  .tm-button {
    padding: 0 $padding-base 0 $padding-base;
  }
}
