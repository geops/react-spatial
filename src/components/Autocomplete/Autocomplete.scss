.tm-autocomplete {
  .tm-search {
    color: $gray-dark;
  }

  .tm-autocomplete-results,
  .tm-search-input {
    border: 2px solid $gray-lighter;
  }

  .tm-bt-clear {
    color: black;
  }

  .tm-autocomplete-results {
    position: absolute;
    margin: auto;
    border-top: none;
    top: calc(100% - 1px);
    left: 0;
    right: 0;
    background-color: white;
    overflow: auto;
    max-height: 400px;
    z-index: $zindex-higher;

    p {
      font-size: 0.9em;
      font-weight: bold;
      color: $gray-light;
      padding: 5px 15px 2px 15px;
      margin: 0;
    }

    ul:first-child {
      margin-top: 8px;
    }

    li {
      &:hover,
      &:focus {
        color: $btn-primary-color;
      }
    }
  }
}
