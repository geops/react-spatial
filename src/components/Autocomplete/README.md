#

This demonstrates the use of Autocomplete.

```jsx
import React from  'react'
import Autocomplete from 'react-spatial/components/Autocomplete';

class AutocompleteExample extends React.PureComponent {
  constructor(props) {
    super(props);
    this.items = ['New York', 'New England', 'Old England', 'Old York'];
    this.state = {
      value: '',
      suggestions: [...this.items],
    };
  }

  updateSuggestions(value) {
    if (!value) {
      this.setState({ suggestions: this.items, value });
      return;
    }

    this.setState({
      suggestions: this.items.filter(item => item.indexOf(value) !== -1),
      value,
    });
  }

  render() {
    const { value, suggestions } = this.state;
    return (
      <div className="tm-autocomplete-example">
        <Autocomplete
          value={value}
          items={suggestions}
          placeholder="Search ..."
          onChange={val => this.updateSuggestions(val)}
          onSelect={val => val && this.setState({ value: val })}
        />
      </div>
    );
  }
}

<AutocompleteExample />;
```
