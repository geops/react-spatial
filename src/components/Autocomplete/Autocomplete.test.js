import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Autocomplete from './Autocomplete';

configure({ adapter: new Adapter() });

const defaultItems = [
  {
    label: 'foo',
  },
  {
    label: 'bar',
  },
  {
    label: 'foo2',
  },
];

const items = [
  {
    label: 'qux',
  },
  {
    label: 'quux',
  },
  {
    label: 'corge',
  },
];

const state = {
  showList: false,
};

const mountDflt = () => {
  return mount(
    <Autocomplete
      value="fooval"
      defaultItems={defaultItems}
      items={items}
      renderItem={item => item.label}
      getItemKey={item => item.label}
    />,
  );
};

describe('Autocomplete', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('matches snapshot', () => {
      const component = renderer.create(<Autocomplete />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when properties are set', () => {
    test('should match snapshot without items and defaultItems using defaultProps', () => {
      const component = renderer.create(<Autocomplete value="fooval" />);
      state.showList = true;
      component.root.instance.setState(state);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should match snapshot without items and defaultItems', () => {
      const component = renderer.create(
        <Autocomplete
          value="fooval"
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
        />,
      );
      state.showList = true;
      component.root.instance.setState(state);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should match snapshot without defaultItems', () => {
      const component = renderer.create(
        <Autocomplete
          value="fooval"
          items={items}
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
        />,
      );
      state.showList = true;
      component.root.instance.setState(state);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should match snapshot without items', () => {
      const component = renderer.create(
        <Autocomplete
          value="fooval"
          defaultItems={defaultItems}
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      state.showList = true;
      component.root.instance.setState(state);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should match snapshot', () => {
      const component = renderer.create(
        <Autocomplete
          value="fooval"
          defaultItems={defaultItems}
          items={items}
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      component.root.instance.setState(state);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should match snapshot showing list', () => {
      const component = renderer.create(
        <Autocomplete
          value="fooval"
          defaultItems={defaultItems}
          items={items}
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      state.showList = true;
      component.root.instance.setState(state);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    describe('#onDocClick()', () => {
      test('hide the list and removes the focus.', () => {
        const component = mountDflt();
        component.setState({
          showList: true,
          focus: true,
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);

        // Hide the list on click on document
        document.dispatchEvent(new Event('click'));
        expect(component.state('showList')).toBe(false);
        expect(component.state('focus')).toBe(false);
      });

      test('does nothing when click comes from ana element of Autocomplete.', () => {
        const component = mountDflt();
        component.setState({
          showList: true,
          focus: true,
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
        component.instance().onDocClick({
          target: component
            .find('div')
            .at(0)
            .getDOMNode(),
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
      });
    });

    describe('#onFocus()', () => {
      test('updates set showList state property to true', () => {
        const component = mountDflt();
        expect(component.state('showList')).toBe(false);
        expect(component.state('focus')).toBe(false);
        component.find('input').simulate('focus', { target: { value: '' } });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);

        // Hide the list on click on document
        document.dispatchEvent(new Event('click'));
        expect(component.state('showList')).toBe(false);

        // Verifies the listener is remove on unmount
        component.instance().onFocus({ target: { value: '' } });
        expect(component.state('showList')).toBe(true);
        component.instance().componentWillUnmount();
        document.dispatchEvent(new Event('click'));
        expect(component.state('showList')).toBe(true);
      });
    });

    describe('#onBlurInput()', () => {
      test('does nothing if no lastKeyPress parameter', () => {
        const component = mountDflt();
        component.setState({
          showList: true,
          focus: true,
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
        component.find('input').simulate('blur');
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
      });

      test('hide list and renove focus if lastKeyPress is not an arrow down.', () => {
        const component = mountDflt();
        component.setState({
          showList: true,
          focus: true,
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
        component.instance().onBlurInput({}, { which: 41 });
        expect(component.state('showList')).toBe(false);
        expect(component.state('focus')).toBe(false);
      });
    });

    describe('#onKeyPress()', () => {
      test('gives focus to the first element of the list', () => {
        const component = mountDflt();
        component.find('input').simulate('keyup', { which: 40 });
        const li = component.find('li').at(0);
        expect(li.getDOMNode()).toBe(document.activeElement);
      });

      test("doesn't give focus to the first element of the list", () => {
        const component = mountDflt();
        component.find('input').simulate('keyup', { which: 41 });
        const li = component.find('li').at(0);
        expect(li.getDOMNode()).not.toBe(document.activeElement);
      });
    });

    describe('#onKeyPressItem()', () => {
      test('does nothing when another key than arrows is pressed.', () => {
        const component = mountDflt();
        const li = component.find('li').at(1);
        li.getDOMNode().focus();
        li.simulate('keydown', { which: 39 });
        expect(li.getDOMNode()).toBe(document.activeElement);
      });

      test('moves focus to search input.', () => {
        const component = mountDflt();
        const li = component.find('li').at(0);
        li.getDOMNode().focus();
        li.simulate('keydown', { which: 38 });
        expect(component.find('input').getDOMNode()).toBe(
          document.activeElement,
        );
      });

      test('moves focus to previous item.', () => {
        const component = mountDflt();
        const lis = component.find('li');
        const li = lis.at(1);
        li.getDOMNode().focus();
        li.simulate('keydown', { which: 38 });
        expect(lis.at(0).getDOMNode()).toBe(document.activeElement);
      });

      test('moves focus to next item.', () => {
        const component = mountDflt();
        const lis = component.find('li');
        const li = lis.at(1);
        li.getDOMNode().focus();
        li.simulate('keydown', { which: 40 });
        expect(lis.at(2).getDOMNode()).toBe(document.activeElement);
      });

      test('moves focus to default items.', () => {
        const component = mountDflt();
        const lis = component.find('li');
        const li = lis.at(2);
        li.getDOMNode().focus();
        li.simulate('keydown', { which: 40 });
        expect(lis.at(3).getDOMNode()).toBe(document.activeElement);
      });

      test('moves focus to the beginning of the list.', () => {
        const component = mountDflt();
        const lis = component.find('li');
        const li = lis.at(5);
        li.getDOMNode().focus();
        li.simulate('keydown', { which: 40 });
        expect(lis.at(0).getDOMNode()).toBe(document.activeElement);
      });
    });

    describe('#onSelect()', () => {
      test('hide list and remove focus when select an item.', () => {
        const fn = jest.fn();
        const component = mount(
          <Autocomplete
            value="fooval"
            defaultItems={defaultItems}
            items={items}
            renderItem={item => item.label}
            getItemKey={item => item.label}
            onSelect={fn}
          />,
        );
        component.setState({
          showList: true,
          focus: true,
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
        component
          .find('li')
          .at(0)
          .simulate('click', {});
        expect(component.state('showList')).toBe(false);
        expect(component.state('focus')).toBe(false);
        expect(fn).toBeCalledTimes(1);
      });

      test('hide list and remove focus when select on default item.', () => {
        const fn = jest.fn();
        const component = mount(
          <Autocomplete
            value="fooval"
            defaultItems={defaultItems}
            items={items}
            renderItem={item => item.label}
            getItemKey={item => item.label}
            onSelect={fn}
          />,
        );
        component.setState({
          showList: true,
          focus: true,
        });
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
        component
          .find('li')
          .at(4)
          .simulate('click', {});
        expect(component.state('showList')).toBe(false);
        expect(component.state('focus')).toBe(false);
        expect(fn).toBeCalledTimes(1);
      });
    });

    describe('#onChange()', () => {
      test('gives focus to the input and display the list.', () => {
        const fn = jest.fn();
        const component = mount(
          <Autocomplete
            value="fooval"
            defaultItems={defaultItems}
            items={items}
            renderItem={item => item.label}
            getItemKey={item => item.label}
            onChange={fn}
          />,
        );
        expect(component.state('showList')).toBe(false);
        expect(component.state('focus')).toBe(false);
        component.find('input').simulate('change', {});
        expect(component.state('showList')).toBe(true);
        expect(component.state('focus')).toBe(true);
        expect(fn).toBeCalledTimes(1);
      });
    });

    test('hides the list on click on search button if the list is open.', () => {
      const component = mountDflt();
      component.setState({
        showList: true,
        focus: true,
      });
      expect(component.state('showList')).toBe(true);
      expect(component.state('focus')).toBe(true);
      component
        .find('[role="button"]')
        .at(1)
        .simulate('click', {});
      expect(component.state('showList')).toBe(false);
      expect(component.state('focus')).toBe(false);
    });
  });
});
