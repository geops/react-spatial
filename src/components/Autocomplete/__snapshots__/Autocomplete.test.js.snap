// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete when no properties are set matches snapshot 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value=""
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "display": "none",
        }
      }
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={
      Object {
        "display": "none",
      }
    }
  />
</div>
`;

exports[`Autocomplete when properties are set should match snapshot 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value="fooval"
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={null}
  >
    <ul
      className="tm-list"
      role="menu"
    >
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        qux
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        quux
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        corge
      </li>
    </ul>
    <hr />
    my_foo_title
    <ul
      className="tm-list"
      role="menu"
    >
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        foo
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        bar
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot showing list 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value="fooval"
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={null}
  >
    <ul
      className="tm-list"
      role="menu"
    >
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        qux
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        quux
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        corge
      </li>
    </ul>
    <hr />
    my_foo_title
    <ul
      className="tm-list"
      role="menu"
    >
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        foo
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        bar
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without defaultItems 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value="fooval"
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={null}
  >
    <ul
      className="tm-list"
      role="menu"
    >
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        qux
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        quux
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        corge
      </li>
    </ul>
    my_foo_title
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without items 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value="fooval"
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={null}
  >
    my_foo_title
    <ul
      className="tm-list"
      role="menu"
    >
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        foo
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        bar
      </li>
      <li
        className="tm-list-item"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        role="menuitem"
        tabIndex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without items and defaultItems 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value="fooval"
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={
      Object {
        "display": "none",
      }
    }
  >
    my_foo_title
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without items and defaultItems using defaultProps 1`] = `
<div
  className="tm-autocomplete"
>
  <div
    className="tm-search-input"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder=""
      tabIndex="0"
      title=""
      type="text"
      value="fooval"
    />
    <div
      aria-label=""
      className="tm-button tm-bt-clear"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      aria-label=""
      className="tm-button tm-bt-search"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title=""
    >
      submit
    </div>
  </div>
  <div
    className="tm-autocomplete-results"
    style={
      Object {
        "display": "none",
      }
    }
  />
</div>
`;
